File: lib/src/features/auth/data/models/auth_code.dart
import 'dart:convert';

/// Enum representing the type of authentication code
enum AuthCodeType {
  emailConfirmation('email_confirmation'),
  passwordReset('password_reset');

  const AuthCodeType(this.value);
  final String value;

  static AuthCodeType fromString(String value) {
    switch (value) {
      case 'email_confirmation':
        return AuthCodeType.emailConfirmation;
      case 'password_reset':
        return AuthCodeType.passwordReset;
      default:
        throw ArgumentError('Invalid AuthCodeType: $value');
    }
  }
}

/// Data model for authentication codes used in email confirmation and password reset
class AuthCode {
  final String id;
  final String code;
  final String userId;
  final AuthCodeType type;
  final DateTime createdAt;
  final DateTime expiresAt;
  final bool isUsed;
  final DateTime? usedAt;

  const AuthCode({
    required this.id,
    required this.code,
    required this.userId,
    required this.type,
    required this.createdAt,
    required this.expiresAt,
    required this.isUsed,
    this.usedAt,
  });

  /// Creates an AuthCode from a database row map
  factory AuthCode.fromMap(Map<String, dynamic> map) {
    return AuthCode(
      id: map['id'] as String,
      code: map['code'] as String,
      userId: map['user_id'] as String,
      type: AuthCodeType.fromString(map['type'] as String),
      createdAt: DateTime.parse(map['created_at'] as String),
      expiresAt: DateTime.parse(map['expires_at'] as String),
      isUsed: map['is_used'] as bool,
      usedAt: map['used_at'] != null 
          ? DateTime.parse(map['used_at'] as String) 
          : null,
    );
  }

  /// Converts the AuthCode to a map for database insertion
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'code': code,
      'user_id': userId,
      'type': type.value,
      'created_at': createdAt.toIso8601String(),
      'expires_at': expiresAt.toIso8601String(),
      'is_used': isUsed,
      'used_at': usedAt?.toIso8601String(),
    };
  }

  /// Creates an AuthCode from JSON string
  factory AuthCode.fromJson(String source) {
    return AuthCode.fromMap(json.decode(source) as Map<String, dynamic>);
  }

  /// Converts the AuthCode to JSON string
  String toJson() => json.encode(toMap());

  /// Checks if the authentication code has expired
  bool get isExpired => DateTime.now().isAfter(expiresAt);

  /// Checks if the authentication code is valid (not used and not expired)
  bool get isValid => !isUsed && !isExpired;

  /// Creates a copy of this AuthCode with the given fields replaced
  AuthCode copyWith({
    String? id,
    String? code,
    String? userId,
    AuthCodeType? type,
    DateTime? createdAt,
    DateTime? expiresAt,
    bool? isUsed,
    DateTime? usedAt,
  }) {
    return AuthCode(
      id: id ?? this.id,
      code: code ?? this.code,
      userId: userId ?? this.userId,
      type: type ?? this.type,
      createdAt: createdAt ?? this.createdAt,
      expiresAt: expiresAt ?? this.expiresAt,
      isUsed: isUsed ?? this.isUsed,
      usedAt: usedAt ?? this.usedAt,
    );
  }

  @override
  String toString() {
    return 'AuthCode(id: $id, code: $code, userId: $userId, type: $type, '
           'createdAt: $createdAt, expiresAt: $expiresAt, isUsed: $isUsed, '
           'usedAt: $usedAt)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    
    return other is AuthCode &&
      other.id == id &&
      other.code == code &&
      other.userId == userId &&
      other.type == type &&
      other.createdAt == createdAt &&
      other.expiresAt == expiresAt &&
      other.isUsed == isUsed &&
      other.usedAt == usedAt;
  }

  @override
  int get hashCode {
    return id.hashCode ^
      code.hashCode ^
      userId.hashCode ^
      type.hashCode ^
      createdAt.hashCode ^
      expiresAt.hashCode ^
      isUsed.hashCode ^
      usedAt.hashCode;
  }
}

File: email-service/app/models/auth_code.py
"""Data models for authentication codes."""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, validator


class AuthCodeType(str, Enum):
    """Enum representing the type of authentication code."""
    
    EMAIL_CONFIRMATION = "email_confirmation"
    PASSWORD_RESET = "password_reset"


class AuthCode(BaseModel):
    """Data model for authentication codes used in email confirmation and password reset."""
    
    id: str = Field(..., description="Unique identifier for the authentication code")
    code: str = Field(..., description="Hashed authentication code")
    user_id: str = Field(..., description="ID of the user this code belongs to")
    type: AuthCodeType = Field(..., description="Type of authentication code")
    created_at: datetime = Field(..., description="When the code was created")
    expires_at: datetime = Field(..., description="When the code expires")
    is_used: bool = Field(default=False, description="Whether the code has been used")
    used_at: Optional[datetime] = Field(None, description="When the code was used")
    
    class Config:
        """Pydantic configuration."""
        use_enum_values = True
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }
    
    @validator('user_id')
    def validate_user_id(cls, v):
        """Validate user_id is not empty."""
        if not v or not v.strip():
            raise ValueError("user_id cannot be empty")
        return v
    
    @validator('code')
    def validate_code(cls, v):
        """Validate code is not empty."""
        if not v or not v.strip():
            raise ValueError("code cannot be empty")
        return v
    
    @validator('expires_at')
    def validate_expires_at(cls, v, values):
        """Validate expires_at is after created_at."""
        if 'created_at' in values and v <= values['created_at']:
            raise ValueError("expires_at must be after created_at")
        return v
    
    @property
    def is_expired(self) -> bool:
        """Check if the authentication code has expired."""
        from datetime import timezone
        return datetime.now(timezone.utc) > self.expires_at.replace(tzinfo=timezone.utc)
    
    @property
    def is_valid(self) -> bool:
        """Check if the authentication code is valid (not used and not expired)."""
        return not self.is_used and not self.is_expired
    
    def to_dict(self) -> dict:
        """Convert to dictionary for database operations."""
        return {
            'id': self.id,
            'code': self.code,
            'user_id': self.user_id,
            'type': self.type if isinstance(self.type, str) else self.type.value,
            'created_at': self.created_at.isoformat(),
            'expires_at': self.expires_at.isoformat(),
            'is_used': self.is_used,
            'used_at': self.used_at.isoformat() if self.used_at else None,
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'AuthCode':
        """Create AuthCode from database row dictionary."""
        return cls(
            id=data['id'],
            code=data['code'],
            user_id=data['user_id'],
            type=AuthCodeType(data['type']),
            created_at=datetime.fromisoformat(data['created_at'].replace('Z', '+00:00')),
            expires_at=datetime.fromisoformat(data['expires_at'].replace('Z', '+00:00')),
            is_used=data['is_used'],
            used_at=datetime.fromisoformat(data['used_at'].replace('Z', '+00:00')) if data['used_at'] else None,
        )